### DISCUSSION TITLE
Whats the plan for the e2e tests?

### DISCUSSION THREAD

**Discussion by Jakob-em** [directus/directus#4921] [Archive]
After seeing the amazing setup for the e2e tests that @rijkvanzanten created I would love to work on improving/adding e2e tests to the project. I think good e2e tests would be really beneficial to improve the stability and to keep the high development pace that you guys have.

I am curious what your thoughts are about the next steps related to the e2e tests? What should be added first? Should there also be tests for the admin app? How should the database setup be handled, should there be a fixed sql dump? (maybe that could lead to problems with the different db vendors)

I think running these test in CI is also a big win, so I will setup some GitHub Action for running them.
I would love to hear your thoughts about the topic and also help implementing the tests.

**Comment by rijkvanzanten** [comment]
> I am curious what your thoughts are about the next steps related to the e2e tests?

The next big step is to actually implement the tests (duhhh)

> What should be added first?

I think we should start with some simple "obvious" ones to make sure the whole setup works as expected, like server ping and move on up from there. The system endpoints are easy to test, cause they operate on a known schema

> Should there also be tests for the admin app?

The API is the "engine" for the app as well, so my priority is on having tests for the API first. While the admin app is important to test as well, it's _wayyy_ harder to setup tests for due to the massive amount of _little things_ it does that we'd have to test for

> How should the database setup be handled, should there be a fixed sql dump?

I've been thinking about this as well. At first, I was planning on writing a single seeder that would run on the global setup script, but some tests will inherently alter the database, which in turn means that other tests might fail based on a problem externally. 

As of right now, I'm thinking we should seed (and tear down) the database in the `beforeAll` of the actual tests that are about to run, so we can make sure we're testing the feature set on a known database state. It can _not_ be a SQL dump though, as that won't scale (at all) between the different database vendors. I'm thinking we setup the initial seed data in YAML, and use that to seed the database (similar to how the Directus installer itself works)

> I think running these test in CI is also a big win, so I will setup some GitHub Action for running them.

Absolutely! I spent quite a lot of time making sure the test setup would accept setting specific database vendors and node versions, which in turn means we can use GitHub Actions `matrix` field to run the tests across all databases and node versions simultaneously 😄 

None of this really is set in stone, so I'd love to hear your thoughts as well 🙂

**Reply by Jakob-em** [reply]
Ok starting with the api sounds like a good plan.
I know that e2e tests for UI stuff can get quiet annoying, but I must admit that having the peace of mind that whatever you change doesn't break anything in the UI is really nice 😄 


> As of right now, I'm thinking we should seed (and tear down) the database in the beforeAll of the actual tests that are about to run, so we can make sure we're testing the feature set on a known database state.

That sounds like a good solution for keeping the state of the db clean for every test, but we will probably have problems if we try to parallelise the tests later on. But I also think that for the start this is the way to go

 
>  I'm thinking we setup the initial seed data in YAML, and use that to seed the database (similar to how the Directus installer itself works)

Ok I haven't seen the installer part yet, but I will have a look at it. I hope I can come up with a solution for settings up the db state (based on the yaml format you mentioned) over the weekend. Because most tests won't make any sense without some data in the db.

> which in turn means we can use GitHub Actions matrix field to run the tests across all databases and node versions simultaneously

Funnily enough that's also the first thing that came to my mind when I saw the stuff with the different db vendors in the test setup. Thats the advantage of developing open source, you get unlimited minutes for GitHub Actions haha /s


I hope it's ok to keep this discussion open for further discussion 🙂

**Reply by rijkvanzanten** [reply]
> That sounds like a good solution for keeping the state of the db clean for every test, but we will probably have problems if we try to parallelise the tests later on. But I also think that for the start this is the way to go

By running the tests in different containers per db / node version and by using "scoped" table names for the tests themselves I'm hopeful we can prevent most conflicts there 🤔 

> Ok I haven't seen the installer part yet, but I will have a look at it. I hope I can come up with a solution for settings up the db state (based on the yaml format you mentioned) over the weekend. Because most tests won't make any sense without some data in the db.

Perfect! 

> I hope it's ok to keep this discussion open for further discussion 🙂

Absolutely 👍🏻

**Reply by rijkvanzanten** [reply]
Btw, are you on [Discord](https://directus.chat)? Might be a little easier to discuss things over there if you have any specific questions or problems. (There's a couple Jakobs in the server but I don't know if you're one of them 😅 )
